Rewrite DELETE docs

justindeguzman opened the pull request:
### Changelog category (leave one):
- Documentation (changelog entry is not required)

justindeguzman committed:
Rewrite DELETE docs

robot-clickhouse-ci-2 commented:
<!-- automatic status comment for PR #55856 from ClickHouse/ClickHouse:rewrite-delete-docs -->
*This is an automated comment for commit 757fa750e63c0a96aadf8b38b9b987ec68a42983 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55856/757fa750e63c0a96aadf8b38b9b987ec68a42983/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://github.com/ClickHouse/ClickHouse/blob/master/.github/PULL_REQUEST_TEMPLATE.md?plain=1">❌ failure</a></td></tr>
<tbody>
</table>

thomoco reviewed (commented):
Very nice set of changes, thank you

I was wondering - should we perhaps be recommending wide parts in all use cases where many deletions are expected?

evillique commented on docs/en/sql-reference/statements/delete.md:
````diff
@@ -7,57 +7,81 @@ keywords: [delete]
 ```
 
-`DELETE FROM` removes rows from the table `[db.]table` that match the expression `expr`. The deleted rows are marked as deleted immediately and will be automatically filtered out of all subsequent queries. Cleanup of data happens asynchronously in the background. This feature is only available for the MergeTree table engine family.
+Deleted rows are internally marked as deleted immediately and will be automatically filtered out of all subsequent queries. Cleanup of data happens asynchronously in the background. The `DELETE` statement is only available for the *MergeTree table engine family.
````
Should we specifically say that this is a "lazy" type of deletion and not that the deletion will happen asynchronously in the background, because we will not do anything to actively delete the data but will simply wait until the next merge happens? As a result, the deleted data can be stored for a long time without actually being deleted.

evillique commented on docs/en/sql-reference/statements/delete.md:
```diff
@@ -7,57 +7,81 @@ keywords: [delete]
-`DELETE FROM` removes rows from the table `[db.]table` that match the expression `expr`. The deleted rows are marked as deleted immediately and will be automatically filtered out of all subsequent queries. Cleanup of data happens asynchronously in the background. This feature is only available for the MergeTree table engine family.
+Deleted rows are internally marked as deleted immediately and will be automatically filtered out of all subsequent queries. Cleanup of data happens asynchronously in the background. The `DELETE` statement is only available for the *MergeTree table engine family.
+
+**Do not use the DELETE statement frequently as it can negatively affect ClickHouse performance.**
```
I am also not sure about this statement because yes, it won't really affect the performance as much as it will just take an undefined amount of time to delete the data.

justindeguzman replied:
What's a better way to phrase this?

**Deleting large volumes of data frequently can negatively affect ClickHouse performance.**

Or should we just omit this?

evillique replied:
Maybe something like this?
```
Deleting large volumes of data with the DELETE statement can negatively affect SELECT query performance.
```

evillique commented on docs/en/sql-reference/statements/delete.md:
```diff
@@ -7,57 +7,81 @@ keywords: [delete]
-Lightweight deletes are asynchronous by default. Set `mutations_sync` equal to 1 to wait for one replica to process the statement, and set `mutations_sync` to 2 to wait for all replicas.
+## Handling large `DELETE`s
+
+Large deletes can negatively affect ClickHouse performance. If you are attempting to delete all rows from a table, consider using the [`TRUNCATE TABLE`](/en/sql-reference/statements/truncate) command and then re-create the table.
```
What does `re-create the table` mean in this context? TRUNCATE TABLE deletes all the data in the table but it does not DROP it, so there isn't really anything to recreate

evillique commented on docs/en/sql-reference/statements/delete.md:
```diff
@@ -7,57 +7,81 @@ keywords: [delete]
+
+Large deletes can negatively affect ClickHouse performance. If you are attempting to delete all rows from a table, consider using the [`TRUNCATE TABLE`](/en/sql-reference/statements/truncate) command and then re-create the table.
+
+If you anticipate frequent deletes, consider using a [customer partitioning key](/en/engines/table-engines/mergetree-family/custom-partitioning-key). You can then use the [`ALTER TABLE...DROP PARTITION`](/en/sql-reference/statements/alter/partition#drop-partitionpart) command to quickly drop all rows associated with that partition.
```
```suggestion
If you anticipate frequent deletes, consider using a [custom partitioning key](/en/engines/table-engines/mergetree-family/custom-partitioning-key). You can then use the [`ALTER TABLE...DROP PARTITION`](/en/sql-reference/statements/alter/partition#drop-partitionpart) command to quickly drop all rows associated with that partition.
```

evillique commented on docs/en/sql-reference/statements/delete.md:
```diff
@@ -7,57 +7,81 @@ keywords: [delete]
+
+### `DELETE`s are eventually consistent
+
+`DELETE`s are asynchronous by default. This means that if you use `DELETE` in a multi-replica setup, your rows will not be deleted immediately. To change this behavior, set `mutations_sync` equal to 1 to wait for one replica to process the statement, or set `mutations_sync` to 2 to wait for all replicas.
```
This is not true, the DELETE (marking the rows as deleted) is synchronous by default (mutations_sync = 2), and there is currently nothing except `OPTIMIZE TABLE ... FINAL` that can force ClickHouse to actually synchronously delete the data.

evillique reviewed (changes requested):
We should also compare it to the `ALTER TABLE ... DELETE` in ways that `ALTER DELETE` will actually delete the data, but it will recreate all the affected parts, which may take a lot of resources at the moment of deletion. At the same time, the `DELETE` query will not load ClickHouse as much because we only mark the rows as deleted, but it may take a lot of time for the data to be actually deleted.

justindeguzman committed:
DELETE docs improvements based on feedback

evillique commented on docs/en/sql-reference/statements/delete.md:
````diff
@@ -7,57 +7,79 @@ keywords: [delete]
 ```
 
-`DELETE FROM` removes rows from the table `[db.]table` that match the expression `expr`. The deleted rows are marked as deleted immediately and will be automatically filtered out of all subsequent queries. Cleanup of data happens asynchronously in the background. This feature is only available for the MergeTree table engine family.
+Deleted rows are internally marked as deleted immediately and will be automatically filtered out of all subsequent queries. Cleanup of data happens asynchronously in the background until the next merge. As a result, it is possible that for an unspecified period, data is not actually deleted from storage and is only marked as deleted.
````
This would be more precise
```suggestion
Deleted rows are internally marked as deleted immediately and will be automatically filtered out of all subsequent queries. Cleanup of data happens during the next merge. As a result, it is possible that for an unspecified period, data is not actually deleted from storage and is only marked as deleted.
```

evillique commented on docs/en/sql-reference/statements/delete.md:
````diff
@@ -7,57 +7,79 @@ keywords: [delete]
 DELETE FROM hits WHERE Title LIKE '%hello%';
 ```
 
-Lightweight deletes are asynchronous by default. Set `mutations_sync` equal to 1 to wait for one replica to process the statement, and set `mutations_sync` to 2 to wait for all replicas.
````
Why were all mentions of `DELETE` being lightweight deleted?

justindeguzman replied:
Sorry, I'm not sure I understand this comment. Can you expand further?

evillique replied:
I mean previously DELETEs in the text were explicitly called lightweight, now it is just DELETEs (except line 75)

evillique commented on docs/en/sql-reference/statements/delete.md:
```diff
@@ -7,57 +7,79 @@ keywords: [delete]
 
-`DELETE FROM` removes rows from the table `[db.]table` that match the expression `expr`. The deleted rows are marked as deleted immediately and will be automatically filtered out of all subsequent queries. Cleanup of data happens asynchronously in the background. This feature is only available for the MergeTree table engine family.
+Deleted rows are internally marked as deleted immediately and will be automatically filtered out of all subsequent queries. Cleanup of data happens asynchronously in the background until the next merge. As a result, it is possible that for an unspecified period, data is not actually deleted from storage and is only marked as deleted.
+
```
Should we also somewhere at the top of the page mention and refer people to a "heavy" `ALTER table DELETE` query that will actually give you guarantees (asynchronously or synchronously) of deleting your data in a predictable period of time?

justindeguzman committed:
Docs: edit lightweight DELETE docs based on feedback

justindeguzman commented:
@evillique Thank you for reviewing. I believe I've addressed all of your feedback above in the most recent commit.

Please let me know if there are any more changes I should make.

den-crane commented:
> Very nice set of changes, thank you
> 
> I was wondering - should we perhaps be recommending wide parts in all use cases where many deletions are expected?

May be it has sense to put _row_exists into a separate file **always**, even if a part is compact. Compact parts are introduced to solve the overhead of small inserts, but delete happens when inserts already done and this column does not exists in the data.bin.

evillique reviewed (approved):
Looks perfect, thank you!

evillique committed:
Update aspell-dict.txt

nellicus commented:
thank you @evillique @justindeguzman ! 🙏
